keps/sig-api-machinery/20180731-crd-pruning.md

```diff
@@ -0,0 +1,610 @@
+   {
+     "foo": 42,
+     "json": {
+       "bar": {
+         "inner":43,
+         "abc": 44
+       },
+       "def": 45
+     }
+   }
+   ```
+   
+   is pruned to
+   
+   ```json
+   {
+      "json": {
+        "bar": {
+          "inner": 43
+        },
+        "def": 45
+      }
+   }
+   ```
+   
+9. `additionalProperties` within JSON
+
+   ```yaml
+   type: object
+   properties:
+     json:
+       type: object
+       x-kubernetes-preserve-unknown-fields: true
+       nullable: true
+       additionalProperties:
+         type: object
+   ```
+
+   The `additionalProperties` disables pruning **at its level**, `x-kubernetes-preserve-unknown-fields: true` already has the same effect. Inside of the additional properties `x-kubernetes-preserve-unknown-fields: true` keeps being effective, i.e.
+   
+   ```json
+   {
+     "foo": 42,
+     "json": {
+       "bar": {
+         "inner":43,
+         "abc": 44
+       },
+       "def": 45
+     }
+   }
+   ```
+   
+   is pruned to
+   
+   ```json
+   {
+     "json": {
+       "bar": {
+         "inner":43,
+         "abc": 44
+       },
+       "def": 45
+     }
+   }
+   ```   
+10. embedded resource
+
+   ```yaml
+   type: object
+   properties:
+     object:
+       type: object
+       nullable: true
+       x-kubernetes-embedded-resource: true
+       x-kubernetes-preserve-unknown-fields: true
+   ```
+   
+   Here, inside of `.object` nothing is pruned with the exception of unknown fields under `.object.metadata`, i.e.
+
+   ```json
+   {
+     "foo": 42,
+     "object": {
+       "bar": 43,
+       "abc": 44,
+       "metadata": {
+         "name": "example",
+         "garbage": 45
+       }
+     }
+   }
+   ```
+      
+   is pruned to
+      
+   ```json
+   {
+      "object": {
+        "bar": 43,
+        "abc": 44,
+        "metadata": {
+          "name": "example"
+        }
+      }
+   }
+   ```
+ 
+11. implicit `metav1.TypeMeta` and `metav1.ObjectMeta`
+
+   ```yaml
+   type: object
+   ```
+
+   Pruning takes place, but `apiVersion`, `kind`, `metadata` and known fields under `metadata` are preserved, i.e.
+   
+   ```json
+   {
+     "apiVersion": "example/v1",
+     "kind": "Foo",
+     "metadata": {
+       "name": "example",
+       "garbage": 43
+     },
+     "foo": 42
+   }
+   ```
+   
+   is pruned to
+   
+   ```json
+   {
+     "apiVersion": "example/v1",
+     "kind": "Foo",
+     "metadata": {
+       "name": "example"
+     }
+   }
+   ```
+
+### Opt-in and Opt-out of Pruning on CRD Level
+
+We will add a `PruneUnknownFields` flag to `CustomResourceDefinitionSpec` of `apiextensions.k8s.io/v1beta1`:
+
+```go
+type CustomResourceDefinitionSpec struct {
+  ...
+	
+  // pruneUnknownFields enables pruning of object fields which are not
+  // specified in the OpenAPI schema. apiVersion, kind, metadata and known
+  // fields inside metadata are excluded from pruning.
+  // Defaults to false. 
+  // Setting this field to true is considered an alpha API.
```

Comment:
if this is alpha level, it actually needs to be controlled with a feature gate that is off by default and drops the field data when disabled.

Reply:
fixed. We aim at beta.